lib/ppx/ppx_js.cppo.ml

```diff
@@ -0,0 +1,386 @@
+    | Pcf_val (id, mut, Cfk_concrete (bang, body)) ->
+      let names = check_name id names in
+      names, (`Val (id, mut, bang, body) :: fields)
+    | Pcf_method (id, priv, Cfk_concrete (bang, body)) ->
+      let names = check_name id names in
+      names, (`Meth (id, priv, bang, format_meth self_id body) :: fields)
+    | _ ->
+      Location.raise_errorf ~loc:exp.pcf_loc
+        "This field is not valid inside a js literal object."
+  in
+  try `Fields (snd @@ List.fold_left f (S.empty, []) fields)
+  with Location.Error error -> `Error (extension_of_error error)
+
+
+let literal_object ?loc fields =
+  let obj_lid = random_var () in
+  let constr_lid = random_var () in
+  let self_ty = random_tvar () in
+
+  let fields =
+    List.map (function
+      | `Val (n, mut, bang, body) ->
+        let ty = fresh_type n.loc in
+        `Val (n, mut, bang, body, ty)
+      | `Meth (n, priv, bang, body) ->
+        let rec create_meth_ty exp = match exp.pexp_desc with
+          | Pexp_fun (label,_,_,body) ->
+            (label, fresh_type exp.pexp_loc) :: create_meth_ty body
+          | _ -> []
+        in
+        let ret_ty = fresh_type body.pexp_loc in
+        let fun_ty = create_meth_ty body in
+        `Meth (n, priv, bang, body, (fun_ty, ret_ty))
+    )
+      fields
+  in
+
+  let set_field = function
+    | `Val (id, _, _, body, ty) ->
+      Js.unsafe "set" [ evar obj_lid ; Js.string id.txt ; Exp.constraint_ body ty ]
+        [@metaloc body.pexp_loc]
+    | `Meth (id, _, _, body, (fun_ty, ret_ty)) ->
+      Js.unsafe "set" [
+        evar obj_lid ;
+        Js.string id.txt ;
+        Exp.constraint_
+          (Js.fun_ "wrap_meth_callback" [ (evar id.txt)[@metaloc id.loc] ])
+          (Js.type_ "meth_callback" [
+             Typ.var self_ty ;
+             arrows (List.tl fun_ty) ret_ty
+           ])
+      ]
+  in
+
+  let assignments = sequence (List.map set_field fields) (unit ()) in
+  let meth_def n body (fun_ty,ret_ty) e =
+    [%expr let [%p Pat.var n] = ([%e body] : [%t arrows fun_ty ret_ty]) in [%e e] ] in
+  let meth_defs_and_field_assignments =
+    List.fold_right
+      (fun field e -> match field with
+         | `Val _ -> e
+         | `Meth (n, _, _, body, ty) -> meth_def n body ty e)
+      fields
+      assignments in
+
+  let obj_field_meth_type = function
+    | `Val  (id, _, _, _body, ty) ->
+      (id.txt, [], Js.type_ "prop" [ty])
+    | `Meth (id, _, _, _body, (fun_ty, ret_ty)) ->
+      (id.txt, [], arrows (List.tl fun_ty) (Js.type_ "meth" [ret_ty]))
+  in
+
+  let obj_ty =
+    let l = List.map obj_field_meth_type fields
+    in Typ.object_ l Closed
+  in
+  [%expr
+    let ([%p pvar obj_lid] : [%t Typ.alias (Js.type_ "t" [obj_ty]) self_ty]) =
+      let [%p pvar constr_lid] : [%t Js.type_ "constr" [Typ.var self_ty]] =
+        [%e Js.unsafe "variable" [str "Object"]] in
+      let [%p pvar obj_lid] = [%e Js.unsafe "new_obj" [ evar constr_lid ; Exp.array []]] in
```

Comment:
The generated js will be smaller and easier to read if you use the form: 

```
Js.Unsafe.obj [| "x", Js.Unsafe.inject 10;"y", (Js.Unsafe.inject (wrap_meth_callback ..)) |]
```
It should generate
```
{ x : 5, y : wrap_meth_callback ..}
```
compared to
```
let o = new Object;
o.x = 5;
o.y = wrap_meth_callback  ..;
```